### PR TITLE
syncer: make syncer idempotent

### LINKED ISSUE #83: syncer is not reentrant under multiple unique keys

### LINKED ISSUE #99: make syncer idempotent
In some replay cases, syncer don't keep Idempotency：

- problem 1: if the table don't have primary (even if have unique key), the replay can't keep idempotency.
- problem 2: if we have multiple unique keys, replay can meet key already exist error.

### PR DESCRIPTION
**Pull request by GregoryIan**
fix #99 
fix #83 

- [x] get all unique keys
- [x] make syncer reentrant
- [x] improve concurrent

### CODE REVIEW

**Comment by GregoryIan** [comment]
@iamxy @jomenxiao @holys PTAL

File: syncer/db.go [hunk 1]
```diff
@@ -364,22 +412,29 @@ func genDeleteSQLs(schema string, table string, dataSeq [][]interface{}, columns
...
-		if len(indexColumns) > 0 {
-			whereColumns, whereValues = getColumnData(columns, indexColumns, value)
-		}
-
-		where := genWhere(whereColumns, whereValues)
-		values = append(values, whereValues)
+		ks := getMultipleKeys(columns, value, indexColumns)
 
-		sql := fmt.Sprintf("DELETE FROM `%s`.`%s` WHERE %s LIMIT 1;", schema, table, where)
+		sql, value := getDeleteSQL(schema, table, value, columns, defaultIndexColumn)
```
Comment (andelf) [inline syncer/db.go]: get -> gen

File: syncer/syncer.go [hunk 2]
```diff
@@ -690,7 +686,11 @@ func (s *Syncer) run() (err error) {
 				}
 
 				for i := range sqls {
-					job := newJob(insert, sqls[i], args[i], keys[i], true, pos)
+					key, err = s.resolveKarma(keys[i])
```
Comment (andelf) [inline syncer/syncer.go]: 👍

File: syncer/karma_test.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
+// Copyright 2016 PingCAP, Inc.
```
Comment (andelf) [inline syncer/karma_test.go]: 2017

File: syncer/karma.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+	selectedKarma := keys[0]
+	var nonExistKeys []string
+	for _, key := range keys {
+		if val, ok := k.k[key]; ok {
+			selectedKarma = val
+		} else {
+			nonExistKeys = append(nonExistKeys, key)
+		}
+	}
+	// set karma for those non exist key
```
Comment (andelf) [inline syncer/karma.go]: keys

File: syncer/karma.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
+// Copyright 2016 PingCAP, Inc.
```
Comment (andelf) [inline syncer/karma.go]: 17

File: syncer/syncer.go [hunk 6]
```diff
@@ -677,8 +672,9 @@ func (s *Syncer) run() (err error) {
 			log.Debugf("schema: %s, table %s, RowsEvent data: %v", table.schema, table.name, ev.Rows)
 			var (
 				sqls []string
-				keys []string
+				keys [][]string
 				args [][]interface{}
+				key  string
```
Comment (holys) [inline syncer/syncer.go]: > key declared and not used

Did you compile it and test it ?

File: syncer/config.go [hunk 7]
```diff
@@ -40,6 +40,7 @@ func NewConfig() *Config {
 	fs.StringVar(&cfg.LogRotate, "log-rotate", "day", "log file rotate type, hour/day")
 	fs.BoolVar(&cfg.EnableGTID, "enable-gtid", false, "enable gtid mode")
 	fs.BoolVar(&cfg.AutoFixGTID, "auto-fix-gtid", false, "auto fix gtid while switch mysql master/slave")
+	fs.BoolVar(&safeModel, "safe-mode", false, "enable safe model to make syncer reentrant")
```
Comment (holys) [inline syncer/config.go]: safeModel,  safe-mode ?  Is It safeModel a typo ?

Comment (GregoryIan) [inline syncer/config.go]: yep

File: syncer/syncer.go (hunk 6 above)
Comment (GregoryIan) [inline syncer/syncer.go]: forget remove it，go on your review

**Comment by andelf** [comment]
LGTM

**Comment by holys** [comment]
LGTM

**Pull request merged**